File: DotNetPoly/Statistics.vb
Partial Class GameBoard

    Friend Statistics As GameBoard.StatisticsClass

    Friend Class StatisticsClass
        Private _Gameboard As GameBoard

        Sub New(pGameboard As GameBoard)
            _Gameboard = pGameboard
            ReDim DiceRollSums(pGameboard.Settings.DICE_COUNT * pGameboard.Settings.DICE_EYES)
            ReDim FieldVisits(pGameboard.FieldCount - 1)
            ReDim FieldHovers(pGameboard.FieldCount - 1)
            'ReDim ChanceOccured([Enum].GetNames(GetType(eChanceType)).Length - 1)
            ChanceOccured = New Collections.Hashtable([Enum].GetNames(GetType(eActionType)).Length - 1)
            ActionChoosen = New Collections.Hashtable([Enum].GetNames(GetType(eActionType)).Length - 1)

            'TotalHousefields = pGameboard._Housefields.Values.Count
            For Each field In pGameboard.FieldCol.Values
                If field.GetType() Is GetType(DotNetPoly.Fields.HouseField) Then
                    TotalHousefields += 1
                End If
            Next
            Me.StartTime = DateTime.Now
        End Sub

#Region "Dice"
        Friend TotalDiceCount As Integer
        Friend DiceRollSums() As Integer
        Friend DoublesCount As Integer

        Friend ReadOnly Property TotalEyes As Integer
            Get
                Dim erg As Integer = 0
                For i = 0 To DiceRollSums.GetLength(0) - 1
                    erg += DiceRollSums(i) * i
                Next
                Return erg
            End Get
        End Property

        Friend ReadOnly Property AverageDiceSum As Integer
            Get
                Return CInt(TotalEyes / TotalDiceCount)
            End Get
        End Property
#End Region

        Friend ReadOnly TotalHousefields As Integer
        Friend ReadOnly Property AvailableHouseFields As Integer
            Get
                Return TotalHousefields - SoldHouseFields
            End Get
        End Property
        Friend Property SoldHouseFields As Integer

        Friend CompleteGroups As Integer

        Friend FieldVisits() As Integer
        Friend FieldHovers() As Integer

        Friend ChanceOccured As Collections.Hashtable
        Friend ActionChoosen As Collections.Hashtable

        Friend ReadOnly Property TotalCash As Integer
            Get
                Return _Gameboard.BANK.Cash
            End Get
        End Property

        ReadOnly StartTime As DateTime

        ReadOnly Property PlayTime As TimeSpan
            Get
                Return DateTime.Now - StartTime
            End Get
        End Property

        Friend MaxUpgradeLevel As Integer
        Friend MaxRent As Integer
        Friend MaxCost As Integer

        Friend Lap As Integer


    End Class

End Class

File: DotNetPoly/Statistics.cs
using System;
using System.Linq;

namespace DotNetPoly
{
    partial class GameBoard
    {
        protected internal StatisticsClass Statistics;

        protected internal class StatisticsClass
        {
            private GameBoard _Gameboard;

            public StatisticsClass(GameBoard pGameboard)
            {
                _Gameboard = pGameboard;
                DiceRollSums = new int[pGameboard.Settings.DICE_COUNT * pGameboard.Settings.DICE_EYES + 1];
                FieldVisits = new int[pGameboard.FieldCount - 1 + 1];
                FieldHovers = new int[pGameboard.FieldCount - 1 + 1];
                // ReDim ChanceOccured([Enum].GetNames(GetType(eChanceType)).Length - 1)

                ChanceOccured = new System.Collections.Hashtable(Enum.GetNames(typeof(eActionType)).Length - 1);
                ActionChoosen = new System.Collections.Hashtable(Enum.GetNames(typeof(eActionType)).Length - 1);

                TotalHousefields = (from f in pGameboard.FieldCol where f.Value is Fields.HouseField select f.Value).Count();

                StartTime = DateTime.Now;
            }

            internal int TotalDiceCount;
            internal int[] DiceRollSums;
            internal int DoublesCount;

            internal int TotalEyes
            {
                get
                {
                    int erg = 0;
                    for (int i = 0; i < DiceRollSums.GetLength(0); i++)
                        erg += DiceRollSums[i] * i;
                    return erg;
                }
            }

            internal int AverageDiceSum
            {
                get
                {
                    return Convert.ToInt32(TotalEyes / (double)TotalDiceCount);
                }
            }

            internal readonly int TotalHousefields;
            internal int AvailableHouseFields => TotalHousefields - SoldHouseFields;

            internal int SoldHouseFields { get; set; }

            internal int CreditHouses { get; set; }

            internal int CompleteGroups;

            internal int[] FieldVisits;
            internal int[] FieldHovers;

            internal System.Collections.Hashtable ChanceOccured;
            internal System.Collections.Hashtable ActionChoosen;

            internal int TotalCash => _Gameboard.BANK.Cash;
  
            private readonly DateTime StartTime;

            public TimeSpan PlayTime => DateTime.Now - StartTime;

            internal int MaxUpgradeLevel;
            internal int MaxRent;
            internal ushort MaxCost;

            internal int Lap;
        }
    }
}
